# Contents

1. Importing libraries and files

2. Data Consistency
    - 2.1 Mixed Data Types
    - 2.2 Missing Values
    - 2.2 Duplicates
 
3. Exporting df_prods_clean_no_dups Dataframe
   
4. df_ords Data Cleaning
    - 3.1 Mixed Data Types
    - 3.2 Missing Values
    - 3.3 Duplicates
   
3. Exporting entire Dataframe

# 1. Importing Libraries and Files

In [5]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [6]:
path= r'C:\Users\spada\OneDrive\Data Analytics\02-2023 Instacart Basket Analysis'

In [7]:
df_ords= pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [8]:
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 2. Data Consistency

## 2.1 Mixed Data Types

In [9]:
# create a dataframe
df_test = pd.DataFrame()

In [10]:
#create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [11]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [12]:
#check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [13]:
df_test['mix'] = df_test['mix'].astype('str')

## 2.2 Missing Values

In [14]:
#find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
#View missing values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
#Compare number of rows within the current dataframe, compared to the number in the subset once missing rows are removed
df_prods.shape

(49693, 5)

In [18]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods_clean.shape

(49677, 5)

## 2.3 Duplicates

In [20]:
#Find duplicate rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
#number of rows in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [23]:
#creating new dataframe that does not include duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
# number of rows in df_prods_clean_no_dups
df_prods_clean_no_dups.shape

(49672, 5)

# 3. Exporting df_prods_clean_no_dups Dataframe

In [25]:
# Export df_prods_clean_no_dups dataframe
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 4. df_ords data cleaning

In [26]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

## 4.1 Mixed Data Types

In [27]:
#check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
        print (col, ' mixed')
  else: print(col, ' consistent')

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
eval_set  consistent
order_number  consistent
order_day_of_week  consistent
order_hour_of_day  consistent
days_since_prior_order  consistent


##### No mixed data types found

## 4.2 Missing Values

In [28]:
#find missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [66]:
#Compare number of rows within the current dataframe
df_ords.shape

(3421083, 8)

##### It appears that days_since_prior_order is missing 206209 values. These missing values could be what shows when it is a customer's first ever order.

In [67]:
#View missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [71]:
df_ords_nan = df_ords

In [78]:
#Replacing all null values with -1 to account for customers with first time orders
df_ords_nan['days_since_prior_order'].fillna('-1', inplace = True)

In [73]:
#new dataframe with null values replaced with -1
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         order_day_of_week  order_hour_of_day days_since_prior_order  
0                        2                  8                     -1  
1                        3                  7                   15.0  
2                        3                 12                   21.0  
3                        4                  7                   29.0  
4                        4                 15                   28.0  
...                    ...                ...                    ...  
3421078                  5                 18                   29.0  
3421079                  4                 10                   30.0  
3421080                  1                 12                   18.0  
3421081                  1                 12                    7.0  
3421082                  6                 14                   30.0  

[3421083 rows x 8 columns]

## 4.3 Duplicates

In [75]:
#Find duplicate rows
df_ords_dups = df_ords_nan[df_ords_nan.duplicated()]

In [76]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

##### No duplicates found

In [77]:
#number of rows in df_prods_nan
df_ords_nan.shape

(3421083, 8)

# 5. Exporting entire dataframe

In [80]:
# Export df_ords_nan dataframe
df_ords_nan.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))